### PR TITLE
Payment refactoring

### PR DESCRIPTION
**Pull request by descorp**
# Open PR

## Changes

* split PaymentComponent and PaymentAwareComponent
* move `Payment` to AdyenContext
* simplify `submit ` method
* remove `remainingAmount` from `ComponentManager`
* rename `PaymentAwareComponent` -> `PaymentAware`
* simplify ApplePayPayment struct
* rename `PaymentComponentData.amount` -> `PaymentComponentData.amountToPay`

### CODE REVIEW

File: AdyenCard/Components/Card/CardViewControllerItemsProvider.swift [hunk 1]
```diff
@@ -30,7 +30,7 @@ extension CardViewController {
         private let addressViewModelBuilder: AddressViewModelBuilder
 
         internal init(formStyle: FormComponentStyle,
-                      payment: Payment?,
+                      payment: Payment? = nil,
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Card/CardViewControllerItemsProvider.swift]: let's not use default nil values, because we don't want to miss passing it by mistake :- )

File: AdyenCard/Components/Card/CardViewController.swift [hunk 2]
```diff
@@ -46,7 +46,7 @@ internal class CardViewController: FormViewController {
     internal init(configuration: CardComponent.Configuration,
                   shopperInformation: PrefilledShopperInformation?,
                   formStyle: FormComponentStyle,
-                  payment: Payment?,
+                  payment: Payment? = nil,
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Card/CardViewController.swift]: let's not use default nil values, because we don't want to miss passing it by mistake :- )

File: AdyenCard/Components/Stored Card/StoredCardComponent.swift [hunk 3]
```diff
@@ -27,9 +29,11 @@ internal final class StoredCardComponent: PaymentComponent, PaymentAwareComponen
     private let storedCardPaymentMethod: StoredCardPaymentMethod
     
     internal init(storedCardPaymentMethod: StoredCardPaymentMethod,
+                  payment: Payment? = nil,
```
Comment (mohammedDehairy) [inline AdyenCard/Components/Stored Card/StoredCardComponent.swift]: let's not use default nil values, because we don't want to miss passing it by mistake :- )

File: Adyen/Model/PaymentComponentData.swift [hunk 4]
```diff
@@ -22,6 +22,11 @@ public struct PaymentComponentData {
 
     /// The partial payment order if any.
     public let order: PartialPaymentOrder?
+
+    /// The payment amount.
+    public var amount: Amount? {
+        order?.remainingAmount
+    }
```
Comment (mohammedDehairy) [inline Adyen/Model/PaymentComponentData.swift:31]: lets not removed the amount from `PaymentData`, as we discussed offline :-)

File: AdyenComponents/Boleto/BoletoComponentExtensions.swift [hunk 5]
```diff
@@ -28,14 +30,17 @@ extension BoletoComponent {
         ///   - localizationParameters: Localization parameters.
         ///   - shopperInformation: Pre-filled optional personal information about the shopper
         ///   - showEmailAddress: Indicates whether to show `sendCopyByEmail` checkbox and email text field
-        public init(style: FormComponentStyle,
+        ///   - payment: The payment information.
+        public init(style: FormComponentStyle = FormComponentStyle(),
+                    payment: Payment? = nil,
```
Comment (mohammedDehairy) [inline AdyenComponents/Boleto/BoletoComponentExtensions.swift]: lets not use default nil values, we don't want to encourage  merchant not to provide it :-)

File: Demo/Common/Networking/PaymentsRequest.swift [hunk 6]
```diff
@@ -29,7 +29,7 @@ internal struct PaymentsRequest: APIRequest {
         var container = encoder.container(keyedBy: CodingKeys.self)
 
         let currentConfiguration = ConfigurationConstants.current
-        let amount = data.order?.remainingAmount ?? currentConfiguration.amount
+        let amount = data.amount ?? currentConfiguration.amount
```
Comment (mohammedDehairy) [inline Demo/Common/Networking/PaymentsRequest.swift]: maybe don't let the merchant worry about that and just provide the right amount as discussed offline

File: Tests/AdyenTests/Adyen Tests/Components/Gift Card/GiftCardComponentTests.swift [hunk 7]
```diff
@@ -236,44 +233,6 @@ class GiftCardComponentTests: XCTestCase {
         waitForExpectations(timeout: 10, handler: nil)
     }
 
-    func testMissingPaymentObject() throws {
```
Comment (mohammedDehairy) [inline Tests/AdyenTests/Adyen Tests/Components/Gift Card/GiftCardComponentTests.swift:239]: why is this removed? payment object is still optional, right?

File: AdyenCard/Components/Card/CardViewControllerItemsProvider.swift (hunk 1 above)
Comment (descorp) [reply AdyenCard/Components/Card/CardViewControllerItemsProvider.swift]: In this case it is internal, but sure - good point!

File: AdyenCard/Components/Card/CardViewController.swift (hunk 2 above)
Comment (descorp) [reply AdyenCard/Components/Card/CardViewController.swift]: 👍

File: Adyen/Model/PaymentComponentData.swift (hunk 4 above)
Comment (descorp) [reply Adyen/Model/PaymentComponentData.swift:31]: 👍

File: Demo/Common/Networking/PaymentsRequest.swift (hunk 6 above)
Comment (descorp) [reply Demo/Common/Networking/PaymentsRequest.swift]: hm.. in this case this is not a mobile integration, but a "backend code" 🤔
Not sure how can we do this..

File: Tests/AdyenTests/Adyen Tests/Components/Gift Card/GiftCardComponentTests.swift (hunk 7 above)
Comment (descorp) [reply Tests/AdyenTests/Adyen Tests/Components/Gift Card/GiftCardComponentTests.swift:239]: no, not any more :)
DropIn will emit console log / assert failure, but component itself just expects the amount (like ApplePay)

Comment (mohammedDehairy) [reply Tests/AdyenTests/Adyen Tests/Components/Gift Card/GiftCardComponentTests.swift:239]: why keep it optional then?

File: Demo/Common/Networking/PaymentsRequest.swift (hunk 6 above)
Comment (mohammedDehairy) [reply Demo/Common/Networking/PaymentsRequest.swift]: don't think I understand what you're saying! we discussed this before, right?

File: Adyen/Model/PaymentComponentData.swift [hunk 8]
```diff
@@ -102,22 +102,34 @@ public struct PaymentComponentData {
...
-                             amount: amount,
                              order: order,
                              storePaymentMethod: storePaymentMethod,
                              browserInfo: browserInfo,
+                             checkoutAttemptId: checkoutAttemptId,
                              installments: installments)
     }
 
     @_spi(AdyenInternal)
     public func replacingAmount(with amount: Amount) -> PaymentComponentData {
```
Comment (erenbesel) [inline Adyen/Model/PaymentComponentData.swift]: If amount is removed, this function becomes redundant I think

Comment (descorp) [reply Adyen/Model/PaymentComponentData.swift]: amount is back :)

File: Tests/AdyenTests/Adyen Tests/Components/Gift Card/GiftCardComponentTests.swift (hunk 7 above)
Comment (descorp) [reply Tests/AdyenTests/Adyen Tests/Components/Gift Card/GiftCardComponentTests.swift:239]: It [looks like this](https://github.com/Adyen/adyen-ios/pull/940/files#diff-516c91ca2ef64f84384962f06bd375a2c91af3415467d52c9129f6cae0fa3decR55-R58):

```swift 
public convenience init(paymentMethod: GiftCardPaymentMethod,
                            context: AdyenContext,
                            amount: Amount,
                            style: FormComponentStyle = FormComponentStyle())
```

File: Adyen/Core/Components/Base/BasicComponentConfiguration.swift [hunk 9]
```diff
@@ -8,21 +8,33 @@ import Foundation
...
-    var style: FormComponentStyle { get }
+
+    /// The payment information.
+    var payment: Payment? { get }
+}
+
+/// Any presentable payment component basic configuration.
+public protocol AnyPresentableComponentConfiguration: AnyBasicComponentConfiguration {
+
+    associatedtype StyleType
```
Comment (mohammedDehairy) [inline Adyen/Core/Components/Base/BasicComponentConfiguration.swift]: I don[t this is a good idea, we can't use `AnyPresentableComponentConfiguration` or any of its derivatives (subprotocols ) in a function parameter or return type or var type, then what's the point?

File: Adyen/Core/Components/Base/BasicComponentConfiguration.swift [hunk 10]
```diff
@@ -8,21 +8,33 @@ import Foundation
 
 /// Any component's most basic configuration.
 public protocol AnyBasicComponentConfiguration: Localizable {
-    
-    /// The form component style.
-    var style: FormComponentStyle { get }
+
+    /// The payment information.
+    var payment: Payment? { get }
```
Comment (mohammedDehairy) [inline Adyen/Core/Components/Base/BasicComponentConfiguration.swift]: maybe rename `PaymentAwareComponent` to `PaymentAware` and make `AnyBasicComponentConfiguration` implement it

File: Adyen/Model/PaymentComponentData.swift [hunk 11]
```diff
@@ -13,6 +13,8 @@ import Foundation
  [API Reference](https://docs.adyen.com/api-explorer/#/CheckoutService/latest/post/payments__example_payments-klarna)
  */
 public struct PaymentComponentData {
+
+    internal let internalAmount: Amount?
```
Comment (mohammedDehairy) [inline Adyen/Model/PaymentComponentData.swift]: maybe `totalAmount` makes more sense?

File: Adyen/Model/PaymentComponentData.swift [hunk 12]
```diff
@@ -24,7 +26,9 @@ public struct PaymentComponentData {
     public let order: PartialPaymentOrder?
 
     /// The payment amount.
-    public let amount: Amount?
+    public var amount: Amount? {
```
Comment (mohammedDehairy) [inline Adyen/Model/PaymentComponentData.swift]: maybe `amountToPay`?

File: Adyen/Core/Components/Base/BasicComponentConfiguration.swift (hunk 10 above)
Comment (descorp) [reply Adyen/Core/Components/Base/BasicComponentConfiguration.swift]: 👍

File: Adyen/Model/PaymentComponentData.swift (hunk 12 above)
Comment (descorp) [reply Adyen/Model/PaymentComponentData.swift]: 👍

File: Adyen/Core/Components/Base/BasicComponentConfiguration.swift (hunk 9 above)
Comment (descorp) [reply Adyen/Core/Components/Base/BasicComponentConfiguration.swift]: > we can't use AnyPresentableComponentConfiguration or any of its derivatives (subprotocols ) in a function parameter or return type or var type

this statement is correct, but I doubt we plan to do any of the things above - hence I don't see it as a counter-argument 🤔

However, I will delete AnyPresentableComponentConfiguration - it does not bring much value.
Main change in this PR - `style: FormComponentStyle` is no longer part of `AnyBasicComponentConfiguration`

File: Adyen/Model/PaymentComponentData.swift (hunk 11 above)
Comment (descorp) [reply Adyen/Model/PaymentComponentData.swift]: Since we introduced `amountToPay` - we can just use `amount` 🤔

File: Adyen/Core/Components/Base/BasicComponentConfiguration.swift (hunk 9 above)
Comment (mohammedDehairy) [reply Adyen/Core/Components/Base/BasicComponentConfiguration.swift]: just try not to use `associatedtype`, thats my point, it doesn't add any value :-)

and for `AnyPresentableComponentConfiguration` why does it not add any value? maybe if you remove the associated type, then it will have value :-D

File: Adyen/Core/Core Protocols/PaymentAwareComponent.swift [hunk 13]
```diff
@@ -7,15 +7,15 @@
 import Foundation
 
 /// Any component with a partial payment property.
-public protocol PartialPaymentOrderAware: Component {
+public protocol PartialPaymentOrderAware: AnyObject {
```
Comment (mohammedDehairy) [inline Adyen/Core/Core Protocols/PaymentAwareComponent.swift]: doesn't have to be `AnyObject`, right?

File: Adyen/Core/Core Protocols/PaymentComponent.swift [hunk 14]
```diff
@@ -14,10 +14,6 @@ public protocol PaymentMethodAware: AnyObject {
 
 /// A component that handles the initial phase of getting payment details to initiate a payment.
 public protocol PaymentComponent: Component, PartialPaymentOrderAware, PaymentMethodAware {
-
-    /// The context object for this component.
-    @_spi(AdyenInternal)
-    var context: AdyenContext { get }
```
Comment (mohammedDehairy) [inline Adyen/Core/Core Protocols/PaymentComponent.swift:20]: why is this removed? any `PaymentComponent` must have a `AdyenContext ` object, right?

Comment (descorp) [reply Adyen/Core/Core Protocols/PaymentComponent.swift:20]: This is duplicate.
`Component` is already `AdyenContextAware`

File: Adyen/Core/Core Protocols/PaymentAwareComponent.swift (hunk 13 above)
Comment (descorp) [reply Adyen/Core/Core Protocols/PaymentAwareComponent.swift]: 👍

File: Adyen/Core/Core Protocols/PaymentComponent.swift [hunk 15]
```diff
@@ -27,23 +23,14 @@ public protocol PaymentComponent: PaymentAwareComponent, PaymentMethodAware {
...
-            updatedData = PaymentComponentData(paymentMethodDetails: data.paymentMethod,
-                                               amount: data.amount,
-                                               order: data.order,
-                                               storePaymentMethod: data.storePaymentMethod,
-                                               browserInfo: data.browserInfo,
-                                               checkoutAttemptId: checkoutAttemptId,
-                                               installments: data.installments)
-
-        }
+        let updatedData = data.replacingCheckoutAttemptID(with: checkoutAttemptId)
```
Comment (erenbesel) [inline Adyen/Core/Core Protocols/PaymentComponent.swift:35]: nice!

File: AdyenComponents/Issuer List/IssuerListComponent.swift [hunk 16]
```diff
@@ -100,7 +106,9 @@ extension IssuerListComponent: TrackableComponent {}
 extension IssuerListComponent {
     
     /// Configuration for Issuer List type components.
-    public struct Configuration {
+    public struct Configuration: AnyBasicComponentConfiguration {
+
+        public var payment: Payment?
```
Comment (erenbesel) [inline AdyenComponents/Issuer List/IssuerListComponent.swift]: I think I missed this part before, why do we add Payment to the config? Doesn't feel like the right place and component already has payment

File: AdyenDropIn/Models/DropInConfiguration.swift [hunk 17]
```diff
@@ -110,9 +113,6 @@ public extension DropInComponent {
         /// Use this property to enforce a custom collection of card types.
         public var allowedCardTypes: [CardType]?
 
-        /// Indicates the card brands excluded from the supported brands.
-        public var excludedCardTypes: Set<CardType> = [.bcmc]
```
Comment (erenbesel) [inline AdyenDropIn/Models/DropInConfiguration.swift:114]: Why is this removed?

Comment (descorp) [reply AdyenDropIn/Models/DropInConfiguration.swift:114]: it was not used, because `excludedCardTypes` is internal on `CardComponent.Configuration`

File: Adyen/Core/Core Protocols/PaymentComponent.swift (hunk 15 above)
Comment (descorp) [reply Adyen/Core/Core Protocols/PaymentComponent.swift:35]: 😎

File: AdyenComponents/Issuer List/IssuerListComponent.swift (hunk 16 above)
Comment (descorp) [reply AdyenComponents/Issuer List/IssuerListComponent.swift]: > component already has payment

This PR was intended to resolve following problems:
- some components can't have optional payment.
- for all components payment must be set before viewController is initiated - this potentially leads to merchant's confusion and mistakes.
- "asynchronous" setting of the payment leads to complicated flows of `didSet { .. }` for embedded components. 

Solution here is to makes `payment` read-only. However, now we need a new way to set the payment :)

IoC sounds like a good alternative + With this approach ApplePay, DropIn and GiftCard aligned with other components - now we always set payment during configuration.

> Doesn't feel like the right place

What are you suggesting?

Comment (descorp) [reply AdyenComponents/Issuer List/IssuerListComponent.swift]: Going to move `payment` to a AdyenContext

File: Adyen/Core/AdyenContext/AdyenContext.swift [hunk 18]
```diff
@@ -24,17 +27,28 @@ public final class AdyenContext {
...
     internal init(apiContext: APIContext,
+                  payment: Payment?,
                   analyticsProvider: AnalyticsProviderProtocol) {
         self.apiContext = apiContext
         self.analyticsProvider = analyticsProvider
+        self.payment = payment
+    }
+
+    @_spi(AdyenInternal)
+    public func updated(with payment: Payment?) -> AdyenContext {
```
Comment (erenbesel) [inline Adyen/Core/AdyenContext/AdyenContext.swift]: Maybe similar to others, `replacingPayment(with payment: Payment?)`

File: AdyenDropIn/Components/PreApplePayComponent.swift [hunk 19]
```diff
@@ -92,15 +92,7 @@ internal final class PreApplePayComponent: PresentableComponent,
...
     internal func didSubmit(_ data: PaymentComponentData, from component: PaymentComponent) {
-        let checkoutAttemptId = component.context.analyticsProvider.checkoutAttemptId
-        let updatedData = PaymentComponentData(paymentMethodDetails: data.paymentMethod,
-                                               amount: data.amount,
-                                               order: data.order,
-                                               storePaymentMethod: data.storePaymentMethod,
-                                               browserInfo: data.browserInfo,
-                                               checkoutAttemptId: checkoutAttemptId,
-                                               installments: data.installments)
-        submit(data: updatedData, component: self)
```
Comment (mohammedDehairy) [inline AdyenDropIn/Components/PreApplePayComponent.swift:103]: its better if you call `submit(data:component:)`, it won't hurt but will be consistent :-)

File: AdyenDropIn/DropInComponent.swift [hunk 20]
```diff
@@ -265,12 +261,6 @@ public final class DropInComponent: NSObject,
         (component as? PreApplePayComponent)?.presentationDelegate = self
         
         component._isDropIn = true
-        
-        if let payment = configuration.payment, let remainingAmount = component.order?.remainingAmount {
-            component.payment = Payment(amount: remainingAmount, countryCode: payment.countryCode)
-        } else {
-            component.payment = configuration.payment
-        }
```
Comment (mohammedDehairy) [inline AdyenDropIn/DropInComponent.swift:273]: where is the code that changes the Payment amount in the AdyenContext in case of partial payment order remaining amount? am I missing something?

File: Adyen/Core/AdyenContext/AdyenContext.swift (hunk 18 above)
Comment (descorp) [reply Adyen/Core/AdyenContext/AdyenContext.swift]: How about:

```swift
replacing(payment: Payment?)
```

File: AdyenDropIn/DropInComponent.swift (hunk 20 above)
Comment (descorp) [reply AdyenDropIn/DropInComponent.swift:273]: It is moved to [ComponentManager](https://github.com/Adyen/adyen-ios/pull/940/files/3d61fcfa7558cbfca94d3410eae0dd859d75eef0#diff-fd23d788c9a3e6dec6b94258dbabae37abd2de363812a97a822efe48678d81e3R48-R54)

File: AdyenDropIn/Utilities/ComponentManager.swift [hunk 21]
```diff
@@ -34,26 +36,30 @@ internal final class ComponentManager {
...
         self.order = order
         self.supportsEditingStoredPaymentMethods = supportsEditingStoredPaymentMethods
         self.presentationDelegate = presentationDelegate
+
+        if let payment = context.payment, let remainingAmount = order?.remainingAmount {
+            let payment = Payment(amount: remainingAmount, countryCode: payment.countryCode)
+            self.context = context.updated(with: payment)
+        } else {
+            self.context = context
+        }
```
Comment (mohammedDehairy) [inline AdyenDropIn/Utilities/ComponentManager.swift]: I think there is a potential issue here, that the ComponentManager. context will be different than DropInComponent.context, so there is a risk of bugs here.

why not make the `AdyenContext.payment`  `private(set) var`  and in the `updated(with: payment)` just update the payment instead of creating a new instance :-) (AyenContext is a class)

File: AdyenDropIn/DropInComponent.swift (hunk 20 above)
Comment (mohammedDehairy) [reply AdyenDropIn/DropInComponent.swift:273]: [will continue the discussion here](https://github.com/Adyen/adyen-ios/pull/940#discussion_r919929556) :-)

File: AdyenDropIn/Components/PreApplePayComponent.swift (hunk 19 above)
Comment (descorp) [reply AdyenDropIn/Components/PreApplePayComponent.swift:103]: 👍

File: AdyenDropIn/Utilities/ComponentManager.swift (hunk 21 above)
Comment (descorp) [reply AdyenDropIn/Utilities/ComponentManager.swift]: `DropInComponent.context` is holding an original payment and `ComponentManager.context` - the one is in use.
I'd like to keep `DropInComponent.context` untouched because if you opt-out from GiftCard - you need to get back to an original `payment`

Comment (descorp) [reply AdyenDropIn/Utilities/ComponentManager.swift]: Will make `AdyenContext.payment` private(set)

File: AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift [hunk 22]
```diff
@@ -174,8 +174,7 @@ extension ComponentManager: PaymentComponentBuilder {
                                                                    localizationParameters: configuration.localizationParameters)
 
         if let amount = order?.remainingAmount {
-            let localIdentifier = amount.localeIdentifier ?? configuration.localizationParameters?.locale
-            let configuration = applePay.updating(amount: amount, localeIdentifier: localIdentifier)
```
Comment (mohammedDehairy) [inline AdyenDropIn/Utilities/ComponentManager+PaymentComponentBuilder.swift]: sorry why is this removed?

**Review by mohammedDehairy — APPROVED** [review]

File: Adyen/Model/PaymentComponentData.swift [hunk 23]
```diff
@@ -116,8 +121,22 @@ public struct PaymentComponentData {
                              order: order,
                              storePaymentMethod: storePaymentMethod,
                              browserInfo: browserInfo,
+                             checkoutAttemptId: checkoutAttemptId,
                              installments: installments)
     }
+
+    @_spi(AdyenInternal)
+    public func replacingCheckoutAttemptID(with checkoutAttemptId: String?) -> PaymentComponentData {
```
Comment (mohammedDehairy) [inline Adyen/Model/PaymentComponentData.swift:129]: 'replacingCheckoutAttemptID' is old school :-), can it be:

```
updating(checkoutAttemptId: String?) -> PaymentComponentData

or 

replacing(checkoutAttemptId: String?) -> PaymentComponentData
```

**Review by erenbesel — APPROVED** [review]

File: Adyen/Model/PaymentComponentData.swift (hunk 23 above)
Comment (descorp) [reply Adyen/Model/PaymentComponentData.swift:129]: Will do in another PR 😛
